syntax = "proto3";

import "status.proto";
package demo;

enum DataType {
    NULL = 0;
    INT8 = 1;
    INT16 = 2;
    INT32 = 3;
    INT64 = 4;

    STRING = 20;

    BOOL = 30;

    FLOAT = 40;
    HALFFLOAT = 41;
    DOUBLE = 42;

    VECTOR = 100;
    UNKNOWN = 9999;
}

message VectorColumnInfoPB {
    int64 dimension = 1;
    DataType element_type = 2;
}

message ColumnTypePB {
    DataType type = 1;
    VectorColumnInfoPB info = 2;
}

message ColumnSchemaPB {
    uint64 id = 1;
    string name = 2;
    ColumnTypePB type = 3;
}

message CollectionSchemaPB {
    uint64 id = 1;
    string name = 2;
    repeated ColumnSchemaPB columns = 3;
}

message CreateCollectionRequestPB {
    uint64 request_id = 1;
    CollectionSchemaPB schema = 2;
}

message CollectionSchemaResponsePB {
    Status status = 1;
    CollectionSchemaPB schema = 2;
}

message BinaryVectorValuePB {
    repeated int32 value = 1;
}

message FloatVectorValuePB {
    repeated float value = 1;
}

message VectorColumnValuePB {
    oneof value {
        FloatVectorValuePB float_value = 1;
        BinaryVectorValuePB binary_value = 2;
    }
}

message ColumnValuePB {
    oneof value {
        int32 int8_value = 1;
        int32 int16_value = 2;
        int32 int32_value = 3;
        int64 int64_value = 4;
        float float_value = 5;
        double double_value = 6;
        string string_value = 7;
        bool  bool_value = 8;
        VectorColumnValuePB vector_value = 9;
    }
}

message QueryColumnPB {
    string name = 1;
    ColumnValuePB value = 2;
}

message TermQueryPB {
    string field_name = 1;
    repeated ColumnValuePB values = 2;
}

enum CompareOperator {
    LT = 0;
    LTE = 1;
    EQ = 2;
    GT = 3;
    GTE = 4;
    NE = 5;
}

message CompareExprPB {
    CompareOperator operator = 1;
    ColumnValuePB operand = 2;
}

message RangeQueryPB {
    string field_name = 1;
    repeated CompareExprPB operand = 2;
}

message VectorQueryPB {
    string field_name = 1;
    string vector_query_param = 2;
    int64 topk = 3;
    int64 nq = 4;
    float query_boost = 5;
}

message InnerLeafQueryPB {
    oneof query {
        TermQueryPB term_query = 1;
        RangeQueryPB range_query = 2;
        VectorQueryPB vector_query = 3;
    }
}

message LeafQueryPB {
    InnerLeafQueryPB query = 1;
    float boost = 2;
}

message GeneralQueryPB {
    oneof query {
        LeafQueryPB leaf = 1;
        BinaryQueryPB bin = 2;
    }
}

enum QueryRelation {
    R1 = 0;
    R2 = 1;
    R3 = 2;
    R4 = 3;
}

message BinaryQueryPB {
    GeneralQueryPB left_query = 1;
    GeneralQueryPB right_query = 2;
    QueryRelation relation = 3;
    float boost = 4;
}

message QueryRequestPB {
    uint64 request_id = 1;
    GeneralQueryPB query = 2;
}

message QueryResponsePB {
    Status status = 1;
    uint64 hits = 2;
    repeated int64 ids = 3;
    repeated float score = 4;
    repeated float distance = 5;
}

message CommandPB {
    string cmd = 1;
}

message CollectionSchemaListPB {
    Status status = 1;
    repeated CollectionSchemaPB Collection_schema_list = 2;
}

message InsertRequestPB {
    repeated string field_name = 1;
    repeated ColumnValuePB column_value = 2;
}

message CollectionNamePB {
    string Collection_name = 1;
}

message ConfigPB {
    Status status = 1;
    string config = 2;
}

message IndexParamPB {
    Status status = 1;
    string collection_name = 2;
    string index_param = 3;
}

message GeneralResponsePB {
    Status status = 1;
    string result = 2;
}

service DemoService {
    rpc Search(QueryRequestPB) returns (QueryResponsePB) {}
    
    rpc CreateCollection(CreateCollectionRequestPB) returns (CollectionSchemaResponsePB) {}
    rpc ShowCollections(CommandPB) returns (CollectionSchemaListPB) {}
    rpc DropCollection(CollectionNamePB) returns (GeneralResponsePB) {}

    rpc InsertDocs(InsertRequestPB) returns (GeneralResponsePB) {}
    
    rpc CreateIndex(IndexParamPB) returns (GeneralResponsePB) {}
    rpc DropIndex(CollectionNamePB) returns (GeneralResponsePB) {}
    rpc DescribeIndex(CollectionNamePB) returns (IndexParamPB) {}

    rpc GetConfig(CommandPB) returns (ConfigPB) {}
    rpc SetConfig(ConfigPB) returns (Status) {}
}
